### PR TITLE
fix(jellyfin): 增强标题检索并用 Bangumi ID 校验候选

### PR DESCRIPTION
**Pull request by IT-BillDeng**
## 修改动机

#3190 为带季度标题增加了 `Series → Seasons → Episodes` 层级查询，但真实 Jellyfin 媒体库中仍有一些未覆盖的情况：

- Bangumi 标题及其别名不一定与 Jellyfin 的 Series/Season 标题一致；
- Jellyfin 搜索可能先返回模糊候选，真正匹配的季度位于后续别名中；
- 搜索结果可能直接是 Season、Episode 或 Movie，而不一定是 Series；
- 多季度 Series 的标题和 Provider ID 可能属于第一季，导致后续季度被错误排除；
- `SeriesName` 常是系列通用名，将其作为条目名会让后续 MediaSelector 再次过滤掉已经找到的续作剧集。

这些情况会表现为 Jellyfin 中明明存在对应视频，但 Animeko 找不到任何可播放剧集。

## 为什么使用 Bangumi ID 辅助校验

Animeko 本身以 Bangumi 的条目和剧集数据为基准。从使用场景推断，同时使用 Animeko 与 Jellyfin/Emby 管理动画库的用户，很可能也安装了 [Jellyfin Bangumi 元数据插件](https://github.com/kookxiang/jellyfin-plugin-bangumi)。

标准插件会将 Bangumi 条目 ID 写入 Series/Season 的 `ProviderIds["Bangumi"]`，将 Bangumi 剧集 ID 写入 Episode 的同名字段。因此，本修改可以直接利用已有元数据提高候选匹配置信度：

- 标题、季度和集数仍负责召回候选；
- 剧集 ID 一致时，将候选提升为精确匹配；
- 条目 ID 一致时，提高候选置信度并排除明确冲突；
- 缺少 Bangumi ID 时保持未知状态，继续使用标题、季度和集数匹配，不会排除候选。

实现也兼容服务端额外提供的 `BangumiSubject`；标准插件场景则从父级 Season 继承条目 ID。这里没有恢复不稳定的“直接使用 Bangumi ID 查询 Jellyfin”方案，插件也不是使用 Jellyfin 数据源的硬依赖。

## 修改内容

- 依次尝试完整标题、去季度标题和所有已知别名；在找到精确候选前保留可用的模糊候选作为兜底；
- 同时处理 Series、Season、Episode 和 Movie 搜索结果，并在需要时按层级获取季度和剧集；
- 按请求的 `ep`/`sort` 过滤剧集，并使用当前 Season 标题生成媒体属性，避免被通用 `SeriesName` 二次过滤；
- 获取候选的 Provider IDs，用 Bangumi 条目/剧集 ID 提高匹配置信度或排除明确冲突；
- 只为最终候选补取 MediaStreams，失败时仍保留可播放媒体；
- 限制查询数量并关闭不需要的总数统计；
- 增加覆盖别名、层级季度、模糊兜底、Provider ID 冲突、多季度 Series 和缺失 ID 的单元测试。

## 影响范围

`BaseJellyfinMediaSource` 同时被 Jellyfin 和 Emby 数据源复用，因此两者都会获得新的查询行为。其他数据源的搜索逻辑不受影响。

没有安装 Bangumi 插件或媒体条目缺少 Provider ID 时仍走原有标题/季度/集数匹配。实际服务端验证仅覆盖 Jellyfin，尚未使用真实 Emby 服务端验证。本 PR 不涉及播放器清晰度、转码或播放协议选择。

## 验证

- `./gradlew :datasource:jellyfin:desktopTest :datasource:jellyfin:testAndroidHostTest`
- `git diff --check`
- 拆分前的完整候选通过 `./gradlew check`
- 在包含两个配套修复的 macOS 与 Android 候选应用中，使用真实 Jellyfin 媒体库手动验证：
  - 尼古喵喵
  - 吹响吧！上低音号 第三季
  - 无职转生 第三季
  - 从零开始的异世界生活 丧失篇

### 100 条真实媒体库端到端对照

使用固定种子从真实 Jellyfin 动画库抽取 100 个不同 Bangumi 条目（77 个剧集、23 部 Anime 电影），四组代码使用同一份样本清单。电影样本还要求 Jellyfin 片名匹配 Bangumi 正式名或别名，避免把带有错误 Provider ID 的非 Anime 电影计入结果。

测试不启动客户端，直接运行生产链路 `MediaSourceMediaFetcher → DefaultMediaSelector → awaitCompletedAndSelectDefault`；四组统一使用相同的用户名/密码认证适配作为公共基线，只改变 #3244、#3245 是否存在。

| 代码组合 | 正确候选召回 | 正确自动选择 | 已召回但未选择 |
| --- | ---: | ---: | ---: |
| 基线 | 77/100 | 24/100 | 53 |
| 仅 #3244 | 100/100 | 34/100 | 66 |
| 仅 #3245 | 77/100 | 77/100 | 0 |
| #3244 + #3245 | 100/100 | 100/100 | 0 |

- 本 PR 单独将正确候选召回从 77/100 提升至 100/100；在选择器已修复的对照中，端到端正确选择也从 77/100 提升至 100/100。
- 23 部 Anime 电影在修改前后均正确召回；新增的 23 个正确召回全部来自剧集样本（54/77 → 77/77）。
- “仅本 PR”仍有 66 条已召回但未自动选择，说明剩余失败位于 #3245 处理的候选传播时序，而不是 Jellyfin 查询。
- 四组运行均无异常，样本清单 SHA-256：`8cbd9cd225b0d9b68fdf4e8b15816518a1aaed93b02a0d3e782ff61d1a693464`。

## 关联

- Follow-up to #3190
- 配套的 MediaSelector 时序修复：#3245
- 两个 PR 均可独立审查；本 PR 只修改 Jellyfin/Emby 查询与候选校验。

### CODE REVIEW

**Comment by IT-BillDeng** [comment]
@codex review this PR, focusing on correctness regressions in Jellyfin/Emby candidate retrieval, alias and season fallback behavior, Bangumi ProviderId matching and inheritance, false positives/negatives, request bounds, cancellation behavior, and whether the tests adequately cover production edge cases. Reply in Chinese.

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 1]
```diff
@@ -63,96 +73,293 @@ abstract class BaseJellyfinMediaSource(
...
+            .filter(String::isNotBlank)
+            .distinct()) {
+            val parsedSubjectName = parseSubjectName(subjectName)
+            for (searchName in sequenceOf(subjectName, parsedSubjectName.baseName).distinct()) {
+                val searchResults = doSearch(
+                    subjectName = searchName,
+                    fields = FIELD_PROVIDER_IDS,
+                    limit = TITLE_SEARCH_LIMIT,
+                    enableTotalRecordCount = false,
+                ).Items.filter(Item::isSupportedSearchResult)
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt]: 已修复于 `0fa5149f2`。

标题查询现在：

- 向 Jellyfin/Emby 发送 `includeItemTypes=Series,Season,Episode,Movie`；
- 使用 `startIndex` 分页，每页 50 条；
- 最多读取 4 页（200 条），保证请求有界；
- 仍保留客户端类型过滤，以兼容忽略类型参数的服务端。

测试覆盖了正确候选位于第 51 条，以及服务端持续返回满页时最多只请求 4 页。

整体验证已通过：Desktop 与 Android Host 各 13/13；同一份真实 Jellyfin 100 条样本保持正确召回 100/100、正确自动选择 100/100、异常 0。

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 2]
```diff
@@ -63,96 +73,293 @@ abstract class BaseJellyfinMediaSource(
...
+                }
+                if (exactIdMatches.isNotEmpty()) {
+                    return exactIdMatches
+                }
+
+                if (
+                    preferredMatches.isNotEmpty() &&
+                    preferredCandidates.any { it.hasExactTitle(searchName) }
+                ) {
+                    return fallbackMatches.values.sortedByDescending { it.confidence }
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt]: 已修复于 `a1e6d0e81`。

已移除“仅凭精确标题和请求集号就提前返回”的路径。现在只有 Bangumi 剧集 ID 精确匹配时才会提前结束；缺少 ID 证据的标题候选会作为 fallback 保留，同时继续检查后续别名。

新增回归测试：第一个别名命中同名且集号相同、但没有 Provider ID 的错误候选；第二个别名命中带正确 Bangumi 条目/剧集 ID 的候选。最终只返回后者并标记为精确匹配。

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 3]
```diff
@@ -63,96 +73,293 @@ abstract class BaseJellyfinMediaSource(
...
+                    uri = getDownloadUri(Id),
+                ),
+                originalTitle = originalTitle,
+                publishedTime = 0,
+                properties = MediaProperties(
+                    // SeasonName identifies the current season or arc. SeriesName is often the
+                    // first season's generic title and would be rejected by MediaSelector.
+                    subjectName = SeasonName?.takeIf { it.isNotBlank() }
+                        ?: SeriesName?.takeIf { it.isNotBlank() }
+                        ?: query.subjectNameCN,
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt]: 已修复于 `0a0856f18`。

`SeasonName` 现在只有在匹配 `query.subjectNames` 或主中文名时才会写入 `Media.properties.subjectName`；对于 `Season 3`、`第 3 季` 等通用季度名，会回退到请求中的已知条目名，从而避免被生产 MediaSelector 以 `SubjectNameMismatch` 排除。

测试覆盖了 Jellyfin 返回通用 `Season 3`、但 Bangumi 条目名为具体续作标题的情况。

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 4]
```diff
@@ -63,96 +73,293 @@ abstract class BaseJellyfinMediaSource(
...
-                            item.Name,
-                            EpisodeRange.unknownSeason(),
-                        )
-
-                        else -> return@mapNotNull null
+                    TYPE_SEASON -> {
+                        doSearch(
+                            parentId = candidate.Id,
+                            fields = FIELD_PROVIDER_IDS,
+                            enableTotalRecordCount = false,
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt:315]: 已修复于 `4f7795fba`。

现在当标题解析出目标季度时：

- 直接返回的 Season 会按 `IndexNumber` 过滤；
- 直接返回的 Episode 会按 `ParentIndexNumber` 过滤；
- Series 回退展开出的 Episode 也应用相同季度约束；
- 若候选的 Bangumi 条目 ID 已明确匹配，则保留 ID 作为更强证据，避免季度编号元数据异常造成假阴性。

新增了“多个直接 Season”和“多个直接 Episode”两条无 Provider ID 回归测试，确认只保留目标季度。

**Comment by IT-BillDeng** [comment]
@codex review 请复查针对上一轮四条审查意见的修复，确认问题是否已解决。

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 5]
```diff
@@ -63,98 +75,308 @@ abstract class BaseJellyfinMediaSource(
...
+
+                if (searchResults.isEmpty()) {
+                    continue
+                }
+
+                val preferredCandidates = searchResults.preferredCandidatesFor(searchName)
+                val preferredMatches = findRequestedItems(
+                    preferredCandidates,
+                    query,
+                    parsedSubjectName.targetSeason,
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt]: 已修复于 `49cc612d5`，并在 `21901c5a9`、`6a9ab0501` 补齐边界。

`createSubjectSearches` 现在会先解析完整标题和所有别名中的季度；当整组名称只解析出一个目标季度时，该季度会传播给没有季度信息的别名，搜索项按 `(name, targetSeason)` 去重。因此泛标题搜索仍携带正式标题提供的季度约束。

后续过滤也收紧为：显式 `ParentIndexNumber` 冲突不会被通用 `SeasonName` 绕过。对于“一季一个 Series”的媒体库，仅当该 Series 只有一个 Season，并且 Series 标题精确匹配带季查询，或 Series 的 Bangumi subject ID 一致时，才允许内部 Season 1 映射到目标季度；只有通用系列名时不会放宽。

回归测试覆盖“带季正式名 + 无季别名”、冲突季度、通用名单季 Series、按季度拆分的 Series，以及“无职转生 第二季 下半 / 第三季”。Desktop 与 Android Host 均为 25/25；固定 100 条真实样本正确召回 100/100、正确自动选择 100/100、错误 0。

File: datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt [hunk 6]
```diff
@@ -63,98 +75,308 @@ abstract class BaseJellyfinMediaSource(
...
 
-                            if (matchedSeasons.isNotEmpty()) {
-                                matchedSeasons.asFlow().flatMapConcat { season ->
+                val requestedItems: List<RequestedItem> = when (candidate.Type) {
+                    TYPE_SERIES -> {
+                        val seasons = doGetSeasons(seriesId = candidate.Id).Items
+                        val matchedSeasons = if (targetSeason != null) {
+                            seasons.filter { it.IndexNumber == targetSeason }
+                        } else {
+                            seasons
```
[the remark this answers is not in the document]
Comment (IT-BillDeng) [reply datasource/jellyfin/src/commonMain/kotlin/BaseJellyfinMediaSource.kt]: 已修复于 `49cc612d5`，并在 `21901c5a9`、`6a9ab0501` 补齐边界。

Series 展开得到 Season 后，现在会先获取并检查 Season 的 Provider IDs，再决定是否按季度编号排除：Bangumi subject ID 冲突时排除；ID 一致时优先于缺失或不可靠的 `IndexNumber`；ID 一致但 Season 缺少 `IndexNumber` 时，直接通过 `parentId` 查询其剧集。

递归 Episode 候选也会先校验 Bangumi episode ID，再判断季度和集数。episode ID 精确一致时可修正 Jellyfin 中陈旧的集数编号，但不能覆盖冲突的 subject ID。Series 级 subject ID 只作为“一季一个 Series”的正向证据，不会据此误删多季度 Series。

回归测试覆盖错误 Season `IndexNumber`、缺失 `IndexNumber`、标准插件只在父级保存 subject ID、Episode ID 与陈旧集号冲突、subject ID 冲突，以及单季/多季 Series。Desktop 与 Android Host 均为 25/25；固定 100 条真实样本正确召回 100/100、正确自动选择 100/100、错误 0。

**Comment by IT-BillDeng** [comment]
### 补充：当前 review 修复后的真实 Jellyfin 端到端样本

在当前 PR HEAD `6a9ab0501` 上，继续使用原固定清单重跑 100 条真实 Jellyfin 样本，没有重新抽样。

测试直接运行生产链路：

`MediaSourceMediaFetcher → DefaultMediaSelector → awaitCompletedAndSelectDefault`

本次完整链路同时包含配套的 #3245：

- `correctCandidateRecall` 直接衡量本 PR 的查询与候选召回；
- `correctAutoSelection` 衡量候选召回后继续完成自动选择的完整链路，不能全部归因于本 PR。

| 指标 | 结果 |
| --- | ---: |
| 样本总数 | 100 |
| Episode | 77 |
| Anime Movie | 23 |
| 正确候选召回 | 100/100 |
| 正确自动选择 | 100/100 |
| 已正确召回但未选择 | 0 |
| 运行错误 | 0 |

固定样本清单 SHA-256：

`8cbd9cd225b0d9b68fdf4e8b15816518a1aaed93b02a0d3e782ff61d1a693464`

#### 代表样本

| 类型 | Bangumi Subject / Episode | 样本 | 正确召回 | 正确自动选择 |
| --- | --- | --- | ---: | ---: |
| Episode | `501963` / `1704818` | 無職転生Ⅲ ～異世界行ったら本気だす～ / 帰ってきた日常 | ✓ | ✓ |
| Episode | `425998` / `1353854` | Re:ゼロから始める異世界生活 3rd season 襲擊編 / 濁流 | ✓ | ✓ |
| Episode | `510728` / `1388680` | Re:ゼロから始める異世界生活 3rd season 反擊編 / レグルス・コルニアス | ✓ | ✓ |
| Episode | `1606` / `9026` | 涼宮ハルヒの憂鬱 (2009) / エンドレスエイト | ✓ | ✓ |
| Episode | `152091` / `663144` | 響け！ユーフォニアム2 / かぜひきラプソディー | ✓ | ✓ |
| Episode | `283643` / `1296944` | 響け！ユーフォニアム3 / なついろフェルマータ | ✓ | ✓ |
| Movie | `386195` / `1109659` | 特別編 響け！ユーフォニアム～アンサンブルコンテスト～ | ✓ | ✓ |
| Movie | `44693` / `332049` | 劇場版 魔法少女まどか☆マギカ [新編] 叛逆の物語 | ✓ | ✓ |

代表样本覆盖了共享系列基础标题的不同季度、同一季度的不同篇章、年份后缀标题，以及直接 Movie 候选。完整 JSON 保留了每条样本的期望 Jellyfin Item ID、召回候选、实际选择结果与耗时，便于核对。

<details>
<summary>完整 JSON 样本（展开查看）</summary>

```json
{
  "sourceResult": "result-pr3244-isolated-series-compat.json",
  "manifestSha256": "8cbd9cd225b0d9b68fdf4e8b15816518a1aaed93b02a0d3e782ff61d1a693464",
  "variant": "pr3244-isolated-series-compat",
  "sampleCount": 100,
  "summary": {
    "total": 100,
    "anyCandidateRecall": 100,
    "correctCandidateRecall": 100,
    "anyAutoSelection": 100,
    "correctAutoSelection": 100,
    "selectorMissAfterCorrectRecall": 0,
    "errors": 0
  },
  "totalDurationMillis": 24148,
  "representativeSamples": [
    {
      "subjectId": "501963",
      "episodeId": "1704818",
      "title": "無職転生 ～異世界行ったら本気だす～ / 無職転生Ⅲ ～異世界行ったら本気だす～ / 帰ってきた日常",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "135286b1b5f3b1240bd6f97eb1a03fe8",
      "candidateIds": ["135286b1b5f3b1240bd6f97eb1a03fe8"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "135286b1b5f3b1240bd6f97eb1a03fe8",
      "correctlyAutoSelected": true,
      "durationMillis": 284
    },
    {
      "subjectId": "425998",
      "episodeId": "1353854",
      "title": "Re:ゼロから始める異世界生活 / Re:ゼロから始める異世界生活 3rd season 襲擊編 / 濁流",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "b2f1b1b3cd7aff0e54e8051632d5cfcb",
      "candidateIds": ["b2f1b1b3cd7aff0e54e8051632d5cfcb"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "b2f1b1b3cd7aff0e54e8051632d5cfcb",
      "correctlyAutoSelected": true,
      "durationMillis": 254
    },
    {
      "subjectId": "510728",
      "episodeId": "1388680",
      "title": "Re:ゼロから始める異世界生活 / Re:ゼロから始める異世界生活 3rd season 反擊編 / レグルス・コルニアス",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "b04d922b70c372b81b49edf10279784c",
      "candidateIds": ["b04d922b70c372b81b49edf10279784c"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "b04d922b70c372b81b49edf10279784c",
      "correctlyAutoSelected": true,
      "durationMillis": 247
    },
    {
      "subjectId": "1606",
      "episodeId": "9026",
      "title": "涼宮ハルヒの憂鬱 / 涼宮ハルヒの憂鬱 (2009) / エンドレスエイト",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "f49284164cb1e51f71f7e73daa39618a",
      "candidateIds": ["f49284164cb1e51f71f7e73daa39618a"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "f49284164cb1e51f71f7e73daa39618a",
      "correctlyAutoSelected": true,
      "durationMillis": 198
    },
    {
      "subjectId": "152091",
      "episodeId": "663144",
      "title": "響け！ユーフォニアム / 響け！ユーフォニアム2 / かぜひきラプソディー",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "cacdbd29f74ed2eba99838afe7b80b2c",
      "candidateIds": ["cacdbd29f74ed2eba99838afe7b80b2c"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "cacdbd29f74ed2eba99838afe7b80b2c",
      "correctlyAutoSelected": true,
      "durationMillis": 257
    },
    {
      "subjectId": "283643",
      "episodeId": "1296944",
      "title": "響け！ユーフォニアム / 響け！ユーフォニアム3 / なついろフェルマータ",
      "jellyfinType": "Episode",
      "expectedJellyfinItemId": "d248007c263bc7ecbe2c99d2fa0a4421",
      "candidateIds": ["d248007c263bc7ecbe2c99d2fa0a4421"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "d248007c263bc7ecbe2c99d2fa0a4421",
      "correctlyAutoSelected": true,
      "durationMillis": 254
    },
    {
      "subjectId": "386195",
      "episodeId": "1109659",
      "title": "特別編 響け！ユーフォニアム～アンサンブルコンテスト～",
      "jellyfinType": "Movie",
      "expectedJellyfinItemId": "692faecd717f5f5776cf165d839f8bdd",
      "candidateIds": ["692faecd717f5f5776cf165d839f8bdd"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "692faecd717f5f5776cf165d839f8bdd",
      "correctlyAutoSelected": true,
      "durationMillis": 1026
    },
    {
      "subjectId": "44693",
      "episodeId": "332049",
      "title": "劇場版 魔法少女まどか☆マギカ [新編] 叛逆の物語",
      "jellyfinType": "Movie",
      "expectedJellyfinItemId": "aef56e5c12d97931ec437879a06c4403",
      "candidateIds": ["aef56e5c12d97931ec437879a06c4403"],
      "correctCandidateRecalled": true,
      "selectedJellyfinItemId": "aef56e5c12d97931ec437879a06c4403",
      "correctlyAutoSelected": true,
      "durationMillis": 414
    }
  ]
}
```

</details>

原始 JSON 文件 SHA-256：

`2a500241e8f39bd6fdb2959e935a5e72508d1f91834792efcd91ea99d1352504`

相关回归测试：

- `:datasource:jellyfin:desktopTest`：25/25
- `:datasource:jellyfin:testAndroidHostTest`：25/25
- `git diff --check`：通过

**Comment by StageGuard** [comment]
resolve conflicts please

**Comment by IT-BillDeng** [comment]
Conflicts have been resolved.
